### PR TITLE
Remove `jsx-a11y/click-events-have-key-events` and `jsx-a11y/no-static-element-interactions` disables

### PR DESCRIPTION
**Pull request by pksjce**
Closes https://github.com/github/primer/issues/2793

### Problem

In the `treeView` component, the structure looks like

```
li(treeItem) [keydown, focus, blur]
- div(container) [onClick, onAuxClick]
-- div(levelIndicator)
-- div(subtree chevron) [onClick]
-- content
```
Right now we have multiple`onClick` events on `div` tags which are non-interactive elements with no `treeitem` role on them as well. This needs to be managed using the eslint disables as mentioned in the issue title

### Solution

My solution is to convert above to below structure and move all the event handlers to the `treeitem` role `li` tag.
```
li(treeItem) [keydown, focus, blur, onClick, onAuxClick]
- div(container) 
-- div(levelIndicator)
-- div(subtree chevron)
-- content
```

### Rollout strategy

- [x] Patch release
- [ ] Minor release
- [ ] Major release; if selected, include a written rollout or migration plan

### Testing & Reviewing

I have tested this in storybook and it seems to unproblematic.

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Added/updated previews (Storybook)
- [ ] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [x] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: .changeset/beige-schools-invite.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@primer/react": patch
+---
+
+Remove `jsx-a11y/click-events-have-key-events` and `jsx-a11y/no-static-element-interactions` disables
```
Comment (siddharthkp) [inline .changeset/beige-schools-invite.md:5]: If we don't expect any user changes, we can skip changeset :)

File: src/TreeView/TreeView.tsx [hunk 2]
```diff
@@ -463,35 +474,17 @@ const Item = React.forwardRef<HTMLElement, TreeViewItemProps>(
...
             </div>
             {hasSubTree ? (
-              // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
               <div
                 className={clsx(
                   'PRIVATE_TreeView-item-toggle',
                   onSelect && 'PRIVATE_TreeView-item-toggle--hover',
                   level === 1 && 'PRIVATE_TreeView-item-toggle--end',
                 )}
-                onClick={event => {
```
Comment (siddharthkp) [inline src/TreeView/TreeView.tsx:490]: This might still be needed for the subtree handle to work in the controlled story: https://primer-109b1de022-13348165.drafts.github.io/storybook/?path=/story/components-treeview-features--controlled

https://github.com/primer/react/assets/1863771/3afe6a6c-25f2-47ed-8280-b3641a3cb622

But it is strange that it is a div element 🤔 

Maybe the fix is that this should not be a div or have an explicit role instead?

**Review by siddharthkp — CHANGES_REQUESTED** [review]

File: src/TreeView/TreeView.tsx (hunk 2 above)
Comment (pksjce) [reply src/TreeView/TreeView.tsx:490]: Thanks for looking into this. I was not sure what edge case this handled.
Ideally I'd say it needs to have a `role='button'`? But I'm not sure what do we prefer? Do we want `button` element styled to suit the tree or just add the role to the existing element?

Comment (siddharthkp) [reply src/TreeView/TreeView.tsx:490]: > Ideally I'd say it needs to have a role='button'?

That would have been my guess as well, but it looks like that's not the case.

See this [example of treeview from ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/) and the [spec @ericwbailey wrote for our TreeView component](forhttps://github.com/github/primer/blob/main/apis/tree-view-api.md#the-expandcollapse-chevron-toggle)

Comment (ericwbailey) [reply src/TreeView/TreeView.tsx:490]: Thanks for the tag here! Tree nodes are a little weird in that they're interaction actions, which might suggest buttons (that's what I originally led with), but it's more a special kind of content unto itself. 

The tree pattern it's mapped to an operating system File Explorer tree, so that's a good mental model to use here. The entire tree itself is focusable, but then individual nodes are not buttons to allow for easier traversal, as well as a less polluted interactive element list if navigating via an element list of similar technique.

Comment (joshblack) [reply src/TreeView/TreeView.tsx:490]: This structure is definitely tough with `jsx-a11y/no-static-element-interactions` since I think we need a way to distinguish between clicking on the functional label for the treeitem versus both the label and subtree within a `role="treeitem"` 😬 In this case, the container for the label is expecting these events to propagate so that it can do specific behavior when interacting with the label.

In general, this rule feels like it can flag in unexpected ways and the way it suggests or indicates to fix the issue seem to be incorrect 😞 (e.g. suggesting to make the item an interactive role)

Comment (pksjce) [reply src/TreeView/TreeView.tsx:490]: @siddharthkp - Thanks so much for the link! Really enjoyed reading through such a thorough document.

The spec seems to indicate that the chevron icon exists only to be clicked upon (input pointer only). Keyboard and Screenreaders don't need to know about this interaction. If it was a button, it would be an unnecessary burden for Assistive Technology? This is my interpretation. @ericwbailey 

As for the `jsx-a11y/no-static-element-interactions` rule, I feel like I could build a case for it to be ignored or deprioritised on our accessibility scorecard. I would include `jsx-a11y/click-events-have-key-events` in the list too. WDYT @joshblack ?

Comment (ericwbailey) [reply src/TreeView/TreeView.tsx:490]: >If it was a button, it would be an unnecessary burden for Assistive Technology? This is my interpretation. @ericwbailey

Yup, pretty much. The reason behind the approach here is say there's a Treeview with 1,500 files, as well as nested nodes to contain them. Providing buttons here would create 1,500+ tab stops to get through the component.

By making only the Treeview itself focusable and then relying on ARIA announcements and non-<kbd>Tab</kbd> keyboard traversal, we reduce this to one tab stop. This helps out a ton with a lot of different motor control-related accessibility considerations.

Comment (pksjce) [reply src/TreeView/TreeView.tsx]: Ok, so for now I've removed this code change. Elaborated the reasons in the comments and recorded this info in the  issue https://github.com/github/primer/issues/2793

File: src/TreeView/TreeView.tsx [hunk 3]
```diff
@@ -463,23 +474,15 @@ const Item = React.forwardRef<HTMLElement, TreeViewItemProps>(
...
-                onSelect(event)
-              } else {
-                toggle(event)
-              }
-            }}
-            onAuxClick={event => {
-              if (onSelect && event.button === 1) {
-                onSelect(event)
-              }
-            }}
```
Comment (pksjce) [inline src/TreeView/TreeView.tsx:477]: Hey @siddharthkp - How do you feel about this change? I saw that all storybooks worked when i moved these events to the parent.

**Comment by siddharthkp** [comment]
@pksjce The controlled story is a bit broken, i'm not sure if that's an issue in the component or just the story 🤔 

https://github.com/primer/react/assets/1863771/f22603e3-7b3b-4926-892e-a1787c935515

Tested this link: https://primer-109b1de022-13348165.drafts.github.io/storybook/?path=/story/components-treeview-features--controlled

File: src/TreeView/TreeView.tsx (hunk 3 above)
Comment (siddharthkp) [reply src/TreeView/TreeView.tsx:477]: Not confident 😓 https://github.com/primer/react/pull/3853#issuecomment-1794546357

I'm wondering if we're keeping once instance of eslint-ignore of these rules, should we keep the other one as well.

Bonus: This component is very detailed, should we write some tests instead of manual tests to bring confidence?

**Comment by mperrotti** [comment]
> Completely optional/bonus: This component is very detailed, should we write some tests instead of manual tests to increase confidence in the refactor?

Agree with Sid on this one. TreeView is a complex beast, and I'd feel a lot less scared making updates if we had more thorough tests.

File: docs/src/@primer/gatsby-theme-doctocat/live-code-scope.js [hunk 4]
```diff
@@ -7,9 +7,12 @@ import {Placeholder} from '@primer/react/Placeholder'
...
 
-const ReactRouterLink = ({to, ...props}) => {
-  // eslint-disable-next-line jsx-a11y/anchor-has-content
-  return <a href={to} {...props} />
+const ReactRouterLink = ({to, children, ...props}) => {
+  return (
+    <a href={to} {...props}>
+      {children}
+    </a>
+  )
```
[the remark this answers is not in the document]
Comment (pksjce) [reply docs/src/@primer/gatsby-theme-doctocat/live-code-scope.js:15]: Yes, Lets do this. I'll create a new PR for this.

**Comment by pksjce** [comment]
@siddharthkp - You were right about the controlled component subtree selection bug. Thats exactly what this code does 😓 
Upon revert it got fixed. 
This component could definitely use a lot more tests for these edge cases.

**Comment by pksjce** [comment]
@siddharthkp - I made a commit fixing the issue along with the relevant test case. If you remove the `stopPropagation`, you can see that the new test fails. The component does have a ton of tests but it was lacking one for selection of child elements.

**Review by siddharthkp — APPROVED** [review]
Looking good now! 

Just need to remove the debug statements, approving in advance
